### PR TITLE
 Adding SAI PTF tests for NeighborFdbAgeoutTest

### PR DESCRIPTION
**Pull request by chitra-raghavan**
Adding SAI PTF tests for NeighborFdbAgeoutTest
(https://github.com/opencomputeproject/SAI/blob/master/test/saithrift/tests/SAITest.md)

L3IPv4NeighborFdbAgeoutTest
L3IPv6NeighborFdbAgeoutTest

### CODE REVIEW

File: test/saithrift/tests/sail3.py [hunk 1]
```diff
@@ -2025,3 +2025,364 @@ def runTest(self):
...
+            with value SAI_PACKET_ACTION_DROP
+
+        Part1:
+        11. Flush the FDB table entry and re-send the packet from port
+        12. Observe the SAI switch take the action according to the switch attribute.
+
+        Part2:
+        13. Wait for the FDB entry age out time and resend the traffic from Port 2
+        14. Observe the SAI switch take the action according to the switch attribute.
+        15. Restore the Switch attribute to FORWARD and remove the router interface and
```
Comment (atitjain) [inline test/saithrift/tests/sail3.py]: is 15 a teardown step?

File: test/saithrift/tests/sail3.py [hunk 2]
```diff
@@ -2025,3 +2025,364 @@ def runTest(self):
...
+class L3IPv4NeighborFdbAgeoutTest(sai_base_test.ThriftInterfaceDataPlane):
+    def runTest(self):
+        """
+        Description:
+        Create two VLAN router interfaces in same VRF. Create nhop and route for
+        destination. Learn ARP for the nhop and check FDB table for the MAC learnt from
+        the ARP packet in that VLAN. Simulate FDB age-out by clearing the FDB entry or
+        wait for age-out time. After the FDB entry is cleared, the layer 3 traffic via
+        this nhop must take the action as specified by SAI switch attribute
+        SAI_SWITCH_ATTR_FDB_UNICAST_MISS_PACKET_ACTION
```
[the remark this answers is not in the document]
Comment (chitra-raghavan) [reply test/saithrift/tests/sail3.py]: @atitjain , You mean , the switch attribute  "SAI_SWITCH_ATTR_FDB_UNICAST_MISS_PACKET_ACTION" is not supoprted in many ASICs ? In that case is it fine to introduce  if  clause statements for supported platforms ?

[the remark this answers is not in the document]
Comment (atitjain) [reply test/saithrift/tests/sail3.py]: SAI_SWITCH_ATTR_FDB_UNICAST_MISS_PACKET_ACTION is supported in most of the ASIC. 

The assumption in SAI pipeline is that the bridge pipline follows the router so if there is a MAC aging or MAC delete, you will miss FDB (in the egress direction) and the bridge pipeline will handle the FDB miss. This is not something that ASICs support. So all the ASIC can do is drop the packet.

File: test/saithrift/tests/sail3.py (hunk 1 above)
Comment (chitra-raghavan) [reply test/saithrift/tests/sail3.py]: yes. its part of teardown

File: test/saithrift/tests/sail3.py [hunk 3]
```diff
@@ -2025,3 +2025,364 @@ def runTest(self):
...
+
+        Steps:
+        1. Create VLAN 100, 200 in the database
+        2. Associate 2 untagged ports (port 1, 2) as the member port of VLAN 100 and 200 respectively.
+        3. Set port attribute values of "Port VLAN ID 100, 200" for the ports 1 and 2 respectively.
+        4. Create Virtual router V1 and enable V4.
```
[the remark this answers is not in the document]
Comment (chitra-raghavan) [reply test/saithrift/tests/sail3.py]: updated description

File: test/saithrift/tests/sail3.py [hunk 4]
```diff
@@ -2025,3 +2025,364 @@ def runTest(self):
...
+        9. Send traffic on port 2 and observe traffic forwarding to port 1.
+        10. Change the port attribute of port 1 to SAI_SWITCH_ATTR_FDB_UNICAST_MISS_PACKET_ACTION
+            with value SAI_PACKET_ACTION_DROP
+
+        Part1:
+        11. Flush the FDB table entry and re-send the packet from port
+        12. Observer the SAI switch take the action according to the switch attribute.
+
+        Part2:
+        13. Wait for the FDB entry age out time and resend the traffic from Port 2
```
[the remark this answers is not in the document]
Comment (chitra-raghavan) [reply test/saithrift/tests/sail3.py]: updated description. script already handles this.

File: test/saithrift/tests/sail3.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (chitra-raghavan) [reply test/saithrift/tests/sail3.py]: In the script  the attribute "SAI_SWITCH_ATTR_FDB_UNICAST_MISS_PACKET_ACTION"is explicitly  configured with value SAI_PACKET_ACTION_DROP. 
https://github.com/opencomputeproject/SAI/blob/b8499fc1c16dff7a4be992d623b00fce1e51c0fc/test/saithrift/tests/sail3.py#L2149

**Comment by prsunny** [comment]
@chitra-raghavan , can you resolve the conflict?

**Comment by chitra-raghavan** [comment]
closing this Pull request and opened new #961   after  resolving conflicts

**Comment by prsunny** [comment]
@chitra-raghavan , you can actually update the existing PR after resolving conflicts. It would be good for any future reference and review comments.

**Pull request closed without merging**